### PR TITLE
Avoid row promotion in heterogeneous arrays

### LINKED ISSUE #23994: Heterogeneous Array literals promote row by row
This surprised me:

```
julia> [1 "two"; 3im 4.0; 5 6//1]
3×2 Array{Any,2}:
    1           "two"
 0.0+3.0im  4.0+0.0im
   5//1       6//1
```

I would have expected promotion to have been completely disabled since there's a string in there.  Is this the expected behavior?  Is it useful?  Should we change it?

### PR DESCRIPTION
**Pull request by p-zubieta**
Fixes #23994

This is the shameless obvious fix to the issue. I'll leave making `hvcat` more efficient (including optimizing based on inference when possible) for another occasion.

### CODE REVIEW

**Comment by pabloferz** [comment]
Of course the obvious fix doesn't work, so I went the second obvious fix.

Now the problem is that if I leave the methods definitions within the `@testset` it segfaults during testing. @JeffBezanson  Is there a problem using the `eltype` of the `Vararg` arguments as in here?

**Comment by pabloferz** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by mbauman** [comment]
I believe this will fail with

```
julia> [[1 2]; 3 4]
2×2 Array{Int64,2}:
 1  2
 3  4
```

I'm not as familiar with this code; can we just do:

```
hvcat(rows::Tuple{Vararg{Int}}, xs::Number...) = typed_hvcat(promote_typeof(xs...), rows, xs...)
hvcat(rows::Tuple{Vararg{Int}}, as...) = typed_hvcat(promote_eltypeof(as...), rows, as...)
```

**Comment by pabloferz** [comment]
> can we just do:

Yes, I believe something along those lines should work. Let me try again.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/24017#issuecomment-334836957) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/5a41a09381aec67f006dd94fa51647d70d19f1bb/c3cfbf7_vs_50766d7/report.md). cc @ararslan

**Comment by pabloferz** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Review by mbauman — APPROVED** [review]

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/24017#issuecomment-334954403) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/359b54d2a01ff0a2e61b6bd5192a814181174dab/8e8915d_vs_c6899ec/report.md). cc @ararslan

**Comment by pabloferz** [comment]
Nanosoldier regressions seem noise, but just for kicks: @nanosoldier `runbenchmarks("array" || "linalg", vs = ":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/24017#issuecomment-335810036):
```
NanosoldierError: failed to run benchmarks against primary commit: stored type BenchmarkTools.ParametersPreV006 does not match currently loaded type
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/076bec1eb4094dbf66d40ca80c1987694a81ad75/68c2fc0_vs_93a38b7/report.md)
cc @ararslan

**Comment by kshyatt** [comment]
Looks like nanosoldier is down for a bit - do we want to merge anyway?

**Comment by ararslan** [comment]
Someone could manually verify any slowdowns that Nanosoldier noted on its last successful run. Nanosoldier is indeed down for the count due to an outstanding issue with JLD.

**Comment by pabloferz** [comment]
@nanosoldier `runbenchmarks("array" || "linalg", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/24017#issuecomment-344083523) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6570739b2fc09aa3b07df1753889a4e512f53217/e9b3456_vs_4d5ed2e/report.md). cc @ararslan

**Comment by pabloferz** [comment]
I believe the regressions are the same noisy ones observed all over the place. This should be good to go.
